### PR TITLE
Support undo/redo for control points in the Editor

### PR DESCRIPTION
**Pull request by smoogipoo**
Works towards resolving https://github.com/ppy/osu/issues/10314. Now handles control points.

... In the interim until we have binary object comparisons.

### CODE REVIEW

File: osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs [hunk 1]
```diff
@@ -32,61 +34,99 @@ public void Patch(byte[] currentState, byte[] newState)
...
+        private void findChangedIndices(DiffResult result, LegacyDecoder<Beatmap>.Section section, out List<int> removedIndices, out List<int> addedIndices)
+        {
+            removedIndices = new List<int>();
+            addedIndices = new List<int>();
 
             // Find the index of [HitObject] sections. Lines changed prior to this index are ignored.
-            int oldHitObjectsIndex = Array.IndexOf(result.PiecesOld, "[HitObjects]");
-            int newHitObjectsIndex = Array.IndexOf(result.PiecesNew, "[HitObjects]");
+            int oldSectionStartIndex = Array.IndexOf(result.PiecesOld, $"[{section}]");
+            int oldSectionEndIndex = Array.FindIndex(result.PiecesOld, oldSectionStartIndex + 1, s => s.StartsWith(@"[", StringComparison.Ordinal));
```
Comment (turbedi) [inline osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:85]:
```suggestion
            int oldSectionEndIndex = Array.FindIndex(result.PiecesOld, oldSectionStartIndex + 1, s => s.StartsWith('['));
```

File: osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs [hunk 2]
```diff
@@ -32,61 +34,99 @@ public void Patch(byte[] currentState, byte[] newState)
...
+
+            if (oldSectionEndIndex == -1)
+                oldSectionEndIndex = result.PiecesOld.Length;
+
+            int newSectionStartIndex = Array.IndexOf(result.PiecesNew, $"[{section}]");
+            int newSectionEndIndex = Array.FindIndex(result.PiecesNew, newSectionStartIndex + 1, s => s.StartsWith(@"[", StringComparison.Ordinal));
```
Comment (turbedi) [inline osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:91]:
```suggestion
            int newSectionEndIndex = Array.FindIndex(result.PiecesNew, newSectionStartIndex + 1, s => s.StartsWith('['));
```

**Comment by smoogipoo** [comment]
It doesn't seem to be perfect in all cases:
- Changing scroll speed doesn't revert in osu! mode. It _does_ revert in e.g. mania, where it actually has an effect.
- Changing slider velocity in the timeline doesn't revert.

These are issues elsewhere and relate to the way `DifficultyControlPoints` are extracted. It's a little bit of a mess, and this is a best effort.

**Comment by peppy** [comment]
Probably fine / much better than nothing at all. Not sure the scroll speed setting should even be visible there.

**Comment by smoogipoo** [comment]
Sorry for the force push, I changed no earlier code, only removed a new commit that tried to fix `DifficultyControlPoints` not updating on `HitObject`s, then realised it caused damage on the actual timing screen :( Going to require more thought as previously said.

**Comment by bdach** [comment]
Are the assertion failures from the earlier runs here ([example 1](https://github.com/ppy/osu/runs/6856350601?check_suite_focus=true), [example 2](https://github.com/ppy/osu/runs/6856481102?check_suite_focus=true)) related to the removed commit? As in, is this in a reviewable state right now?

**Comment by smoogipoo** [comment]
Nah, I'm currently fixing the assertion failures.

**Comment by smoogipoo** [comment]
Should be reviewable now.

**Comment by bdach** [comment]
Also not sure if related, but there's a weird bug with the bottom timeline display where control points are supposed to be shown:

https://user-images.githubusercontent.com/20418176/173310853-b1c1fcaf-7417-4fde-97ee-62789e820629.mp4

Appears that consecutive undoes hide and show the control point indicators... for some reason? Not sure if related, just bringing it up as something I noticed.

**Comment by smoogipoo** [comment]
The timeline issue could be related to incrementally adding every control point rather than giving a list with every control point populated at once. But probably a separate issue from this branch...?

File: osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs [hunk 3]
```diff
@@ -32,61 +34,105 @@ public void Patch(byte[] currentState, byte[] newState)
...
-            {
-                IBeatmap newBeatmap = readBeatmap(newState);
-                foreach (int i in toAdd)
-                    editorBeatmap.Insert(i, newBeatmap.HitObjects[i]);
-            }
-
-            editorBeatmap.EndChange();
+            // The expected usage of this returned list is to iterate from the start to the end of the list, such that
+            // these indices need to appear in reverse order for the usage to not have to deal with decrementing indices.
+            removedIndices.Reverse();
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:135]: Sure, I was also on the edge about this.

**Review by bdach — DISMISSED** [review]
Looks ok to me

**Comment by peppy** [comment]
This doesn't seem to handle the case where a new type is added to an existing group, is that known / expected?

**Comment by smoogipoo** [comment]
Probably because the type isn't doing anything? It's only undoing/redoing based on what actually gets written to the file, so those points are probably becoming redundant & removed during write?

**Comment by peppy** [comment]
Hmm... that's... very awkward. Especially since it doesn't clear redo states. So you hit redo and suddenly you are in a completely broken looking state.

**Comment by peppy** [comment]
Something has regressed with selected rows on the timing screen. After cloning a point, the newly cloned point is not selected visually (but seems to have the selection internally). I've checked that this works correctly on `master`

https://user-images.githubusercontent.com/191335/173552802-ab04af4e-a401-4be0-a81e-297b0437da7c.mp4

- Select point
- Move to new time
- Clone

Observed row not selected.

**Comment by smoogipoo** [comment]
Admittedly I didn't thoroughly test redoing so I didn't notice it breaking it, though your video doesn't show it.

**Comment by peppy** [comment]
I had a video which has me redoing, but I didn't know what was going on and didn't think it would transfer well to a video 😅

**Comment by smoogipoo** [comment]
Please try this PR again and see if you can break it still. I think I might've resolved both the above issues by not replacing all control points.

There's still some weirdness (which I can't handle), that's best described with a video:
https://user-images.githubusercontent.com/1329837/174539856-fe30be08-2472-43f2-a87b-fb106dd0bbae.mp4

In the video, I'm doing much the same actions as https://github.com/ppy/osu/pull/18668#issuecomment-1154979283.
1. I add an effect point to the first timing point.
    - No undo/redo state created (redundant point).
2. I duplicate an effect point later on.
    - No undo/redo state created (redundant point).
3. I change the duplicated effect point to adjust kiai time.
    - Creates an undo state.
4. I press undo.
    - Both the duplicated effect point and the effect point added in (1) are removed.
5. I press redo.
    - The effect point added in (1) does not return.
    - The duplicated effect point is merged into the one following it (which also starts kiai), because the one following it has been deemed redundant.

**Comment by peppy** [comment]
I can still get in a state where the row isn't selected. I'm not too concerned about certain changes remaining unsupported (like the scroll speed one in the video) but having the row non-selectable is quite breaking. Haven't looked at the code changes yet, but maybe a fix can be applied more locally to the table control rather than adding the overhead with equality etc?

https://user-images.githubusercontent.com/191335/174545301-cf83e153-5156-4991-a78b-a0c9364c4356.mp4

**Comment by smoogipoo** [comment]
Looks like the breaking is reproed by:
1. Select timing point.
2. Adjust BPM.
3. Select any other point.
4. Undo.

> fix can be applied more locally to the table control rather than adding the overhead with equality etc

That's exactly what I was doing before, but it's even worse that way (breakage-wise) because performance starts to become a factor and you need to add _layers_ of workarounds to resolve it.

I prefer equality, but let's see if the above issue changes my view on things.

**Comment by smoogipoo** [comment]
Above issue should be fixed. Though I understand there's quite a lot of files changed here, I still believe that control point equality is the way to go.

I can split the changes out if it makes it easier to review - these changes are MVP and I'd also like to annotate `ControlPointInfo` with #nullable but that's a bit more involved, so I could bundle this in a separate PR.

File: osu.Game/Screens/Edit/Timing/ControlPointTable.cs [hunk 4]
```diff
@@ -44,6 +44,7 @@ public IEnumerable<ControlPointGroup> ControlGroups
                 {
                     BackgroundFlow.Add(new RowBackground(group)
                     {
+                        Selected = group.Equals(selectedGroup?.Value),
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Edit/Timing/ControlPointTable.cs:47]: Nah, will change.

- I previously changed `Equals()` to compare by reference, but didn't include it in the end. I've changed it to include it now.
- Changed this to `ReferenceEquals`

File: osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs [hunk 5]
```diff
@@ -34,61 +37,112 @@ public void Patch(byte[] currentState, byte[] newState)
...
+            processTimingPoints(result, () => newBeatmap ??= readBeatmap(newState));
+            editorBeatmap.EndChange();
+        }
+
+        private void processTimingPoints(DiffResult result, Func<IBeatmap> getNewBeatmap)
+        {
+            findChangedIndices(result, LegacyDecoder<Beatmap>.Section.TimingPoints, out var removedIndices, out var addedIndices);
+
+            if (removedIndices.Count == 0 && addedIndices.Count == 0)
+                return;
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:53]: Main concern was reverting hitobject changes on maps like https://osu.ppy.sh/beatmapsets/65994#osu/203906 that have many control points, but looks okay. The code below could also be changed to use hashset-based prechecks if this becomes an issue.

Removed.

File: osu.Game/Screens/Edit/Timing/ControlPointTable.cs [hunk 6]
```diff
@@ -44,6 +44,7 @@ public IEnumerable<ControlPointGroup> ControlGroups
                 {
                     BackgroundFlow.Add(new RowBackground(group)
                     {
+                        Selected = ReferenceEquals(group, selectedGroup?.Value),
```
Comment (peppy) [inline osu.Game/Screens/Edit/Timing/ControlPointTable.cs:47]: The null check on `selectedGroup` here is causing code quality fail. I'd hope this isn't actually needed, but if it is then moving this to an `updateDisplay` or `Schedule`ing is a better option.

**Review by bdach — APPROVED** [review]

**Review by peppy — APPROVED** [review]
Seems pretty good

**Pull request merged**